### PR TITLE
Feature/security jwt

### PR DESCRIPTION
**Pull request by Elvinasl**
API filter added

### CODE REVIEW

File: src/main/java/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+        // We return it so we can chain more configuration
+        return http;
+    }
+
+    @Bean
+    public CorsConfigurationSource corsConfigurationSource() {
+        CorsConfiguration configuration = new CorsConfiguration();
+        configuration.setAllowedOrigins(Arrays.asList("http://localhost"));
+        configuration.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "DELETE"));
```
Comment (tchristerus) [inline src/main/java/config/SecurityConfig.java]: What about the PATCH method?

File: src/main/java/config/security/filters/ApiValidationFilter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        // If header is present, try grab user principal from database and perform authorization
+        Authentication authentication = getApiAuthentication(request, apiKey);
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+
+        chain.doFilter(request, response);
+    }
+
+    private Authentication getApiAuthentication(HttpServletRequest request, String apiKey) {
```
Comment (tchristerus) [inline src/main/java/config/security/filters/ApiValidationFilter.java]: It should not be possible to get a client by API key. This should return the project that belongs to that API key.

File: src/main/java/config/security/jwt/JwtAuthenticationFilter.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+
+import static com.auth0.jwt.algorithms.Algorithm.HMAC512;
+
+public class JwtAuthenticationFilter extends UsernamePasswordAuthenticationFilter {
+    private AuthenticationManager authenticationManager;
+
+    public JwtAuthenticationFilter(AuthenticationManager authenticationManager) {
```
Comment (tchristerus) [inline src/main/java/config/security/jwt/JwtAuthenticationFilter.java]: The AutoWired attribute is missing here

File: src/main/java/config/security/jwt/JwtAuthenticationFilter.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    public void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) throws IOException, ServletException {
+        // Grab principal
+        ClientPrincipal principal = (ClientPrincipal) authResult.getPrincipal();
+
+        // Create JWT Token
+        String token = JWT.create()
+                .withSubject(principal.getUsername())
+                .withExpiresAt(new Date(System.currentTimeMillis() + JwtProperties.EXPIRATION_TIME))
```
Comment (tchristerus) [inline src/main/java/config/security/jwt/JwtAuthenticationFilter.java]: I am not sure about this one but I think Date should be LocalDate. If I remember right Sander told us that in the classes.

File: src/main/java/config/security/jwt/JwtAuthorizationFilter.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import static com.auth0.jwt.algorithms.Algorithm.HMAC512;
+
+public class JwtAuthorizationFilter extends BasicAuthenticationFilter {
+    private ClientRepository clientRepository;
+
+    public JwtAuthorizationFilter(AuthenticationManager authenticationManager, ClientRepository clientRepository) {
```
Comment (tchristerus) [inline src/main/java/config/security/jwt/JwtAuthorizationFilter.java]: The AutoWired tag is missing here

File: src/main/java/config/security/jwt/JwtProperties.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package config.security.jwt;
+
+class JwtProperties {
```
Comment (tchristerus) [inline src/main/java/config/security/jwt/JwtProperties.java]: SECRET AND EXPIRATION_TIME should come from the properties file

File: src/main/java/controllers/TestsController.java [hunk 7]
```diff
@@ -1,4 +1,27 @@
...
 
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import repositories.ClientRepository;
+
+@RestController
 public class TestsController {
```
Comment (tchristerus) [inline src/main/java/controllers/TestsController.java]: This file could be removed before merging to develop I think

File: src/main/java/config/SecurityConfig.java (hunk 1 above)
Comment (Elvinasl) [reply src/main/java/config/SecurityConfig.java]: We are not using it

File: src/main/java/config/security/jwt/JwtAuthenticationFilter.java (hunk 4 above)
Comment (Elvinasl) [reply src/main/java/config/security/jwt/JwtAuthenticationFilter.java]: Thats library. It asks for Date

File: src/main/java/config/security/jwt/JwtProperties.java (hunk 6 above)
Comment (Elvinasl) [reply src/main/java/config/security/jwt/JwtProperties.java]: Yea, but how can you do that? it's static..

Comment (Elvinasl) [reply src/main/java/config/security/jwt/JwtProperties.java]: Expiration time doesn't really matter. Secret is enough.

**Comment by Elvinasl** [comment]
@alwintje @tchristerus @Zanril95 Lets review this and merge it.

**Pull request merged**